pyiron_atomistics/atomistics/structure/neighbors.py

```diff
@@ -1047,3 +1047,147 @@ def get_cluster(dist_vec, ind_vec, prec=prec):
             ind_shell.append(ia_shells_dict)
         return ind_shell
 
+
+class NeighborsTraj:
+
+    """
+    This class generates the neighbors for a given atomistic trajectory. The assumption here is that the trajectory is
+    canonical (no change in the number and types of species throughout the trajectory). The resulting indices,
+    distances, and vectors are stored as numpy arrays.
+
+    """
+
+    def __init__(self, init_structure=None, positions=None, cells=None, num_neighbors=12, **kwargs):
```

Comment:
I think it'd be extremely useful for this to take any object that implements `HasStructure` instead of passing all of these things separately.  Calling `get_structure` every step might be a bit slower, but since you're doing essentially what `AtomisticGenericJob.get_structure` does anyway, it shouldn't be that much.

Reply:
If it turns out much slower, then I'd say we add a method roughly like this
```python
@classmethod
def from_structures(cls, has_structure):
  structs = list(has_structure.iter_structures())
  return cls(structs[0], [s.positions for s in structs], [s.cell for s in structs])
```
to `NeighborTraj` to achieve the same generality, but offer a small escape hatch for cases where we know `get_structure` is too slow.

Reply:
> If it turns out much slower, then I'd say we add a method roughly like this
> 
> ```python
> @classmethod
> def from_structures(cls, has_structure):
>   structs = list(has_structure.iter_structures())
>   return cls(structs[0], [s.positions for s in structs], [s.cell for s in structs])
> ```
> 
> to `NeighborTraj` to achieve the same generality, but offer a small escape hatch for cases where we know `get_structure` is too slow.

I think this is a great way to generalize this for any derivative of `HasStructure`. However, I don't think building the positions and cells from the individual structures is not that efficient.

So how about something like this:

```
def __init__(self, has_structure=None, init_struct=None, positions=None, cells=None):
    if has_structure is None:
        if positions is None or init_struct is None:
            raise ValueError()

```

and then

```
def _get_neighbors_hs(has_struct, num_neighbors=20, **kwargs):
    [struct.get_neighbors() for struct in has_struct.iter_structures()]
```
which gets called instead of `_get_neighbors`

Reply:
I think we should simply give it a try to use `get_structure()` everywhere and see if it really is slower.  Checking `Trajectory._get_structure` and your code in `_get_neighbors()` looks essentially the same to me, so I don't expect a big hit. 

If you don't want to do it, I can give it a go tonight.

Reply:
Also `__init__` needs to accept an `init` argument as the first argument and pass it to `super().__init__()`, otherwise recursive loading from HDF will not work.

Reply:
> 
> 
> I think we should simply give it a try to use `get_structure()` everywhere and see if it really is slower. Checking `Trajectory._get_structure` and your code in `_get_neighbors()` looks essentially the same to me, so I don't expect a big hit.
> 
> If you don't want to do it, I can give it a go tonight.

Sure I'll merge your PR into this one and see what happens!

Reply:
> 
> 
> Also `__init__` needs to accept an `init` argument as the first argument and pass it to `super().__init__()`, otherwise recursive loading from HDF will not work.

Not sure what you mean by this. Could you clarify?

Reply:
> > Also `__init__` needs to accept an `init` argument as the first argument and pass it to `super().__init__()`, otherwise recursive loading from HDF will not work.
> 
> Not sure what you mean by this. Could you clarify?

When you derive from `DataContainer` the new `__init__` method must be compatible to the original one. One of the ways `DataContainer` is instantiated is by passing a collection to initialize it, like so
```python
d = DataContainer({'a': 1, 'b': 3, 2: 42})
assert d[1] == 3
```
Because this is used by the `DataContainer` internally when loading from hdf, subclasses also need to offer this.  There's a little bit more explanation [here in the attention box](https://pyiron-base.readthedocs.io/en/latest/apidoc/pyiron_base.generic.datacontainer.html?highlight=DataContainer).